src/ui/components/HostPermissions/index.js

```diff
@@ -0,0 +1,151 @@
+/* @flow */
+import * as React from 'react';
+import { compose } from 'redux';
+
+import log from 'core/logger';
+import translate from 'core/i18n/translate';
+import type { I18nType } from 'core/types/i18n';
+import Permission from 'ui/components/Permission';
+
+/* eslint-disable no-continue */
+type Props = {|
+  i18n: I18nType,
+  permissions: Array<string>,
+|};
+
+export class HostPermissionsBase extends React.Component<Props> {
+  constructor(props: Props) {
+    const { i18n } = props;
+    super(props);
+    // These should be kept in sync with Firefox's strings for webextention
+    // host permissions which can be found in
+    // https://hg.mozilla.org/mozilla-central/raw-file/tip/browser/locales/en-US/chrome/browser/browser.properties
+    this.permissionStrings = {
+      allUrls: i18n.gettext('Access your data for all websites'),
+      wildcard: (domain) => {
+        return i18n.sprintf(
+          i18n.gettext('Access your data for sites in the %(domain)s domain'),
+          { domain }
+        );
+      },
+      tooManyWildcards: (count) => {
+        return i18n.sprintf(i18n.ngettext(
+          'Access your data in %(count)s other domain',
+          'Access your data in %(count)s other domains',
+          count), { count: i18n.formatNumber(count) }
+        );
+      },
+      oneSite: (site) => {
+        return i18n.sprintf(
+          i18n.gettext('Access your data for %(site)s'),
+          { site }
+        );
+      },
+      tooManySites: (count) => {
+        return i18n.sprintf(i18n.ngettext(
+          'Access your data on %(count)s other site',
+          'Access your data on %(count)s other sites',
+          count), { count: i18n.formatNumber(count) }
+        );
+      },
+    };
+    this.hostPermissions = [];
+  }
+
+  permissionStrings: Object;
+  hostPermissions: Array<React.Element<typeof Permission>>;
+
+  // Add individual Permission components for individual site/domain permissions.
+  formatItems(items: Array<string>, messageType: string) {
+    for (const item of items) {
+      this.hostPermissions.push(
```

Comment:
I'm finding it hard to run through how things are added to `this.hostPermissions` and why there are added or not because it's a side-effect of a function in some places but directly pushed into in other places.

Is there any reason we have to use `this.hostPermissions.push` rather than just creating an array and returning host permissions in these functions and adding them to an array in `render()` (or in the constructor if that's more efficient).

I don't get why we store them in an array in the object.